pkg/daemon/controller/controller.go

```diff
@@ -252,19 +270,50 @@ func (sc *ServiceController) Reconcile(_ context.Context, req ctrl.Request) (res
 
 	// Reconcile state of Windows services with the ConfigMap data
 	if err = sc.reconcileServices(cmData.Services); err != nil {
+		if failErr := sc.handleReconcileFailure(&node); failErr != nil {
+			klog.Errorf("error handling reconcile error, node may not be cordoned appropriately: %s", failErr)
+		}
 		return ctrl.Result{}, err
 	}
 
 	if err = sc.waitUntilNodeReady(); err != nil {
+		if failErr := sc.handleReconcileFailure(&node); failErr != nil {
+			klog.Errorf("error handling reconcile error, node may not be cordoned appropriately: %s", failErr)
+		}
 		return ctrl.Result{}, fmt.Errorf("error waiting for node to become ready")
 	}
-	// Version annotation is the indicator that the node was fully configured by this version of the services ConfigMap
-	if err = metadata.ApplyVersionAnnotation(sc.ctx, sc.client, node, desiredVersion); err != nil {
-		return ctrl.Result{}, fmt.Errorf("error updating version annotation on node %s: %w", sc.nodeName, err)
+	if err = sc.markNodeConfigured(&node, desiredVersion); err != nil {
+		return ctrl.Result{}, err
 	}
 	return ctrl.Result{}, nil
 }
 
+// handleReconcileFailure should be ran when a reconciliation fails due to issues on the Windows instance
+func (sc *ServiceController) handleReconcileFailure(node *core.Node) error {
+	sc.failureCount += 1
+	if sc.failureCount < 5 {
+		return nil
+	}
+	if err := drain.RunCordonOrUncordon(sc.drainer, node, true); err != nil {
+		return err
+	}
+	return nil
+}
+
+// markNodeConfigured marks the node as configured and schedulable
+func (sc *ServiceController) markNodeConfigured(node *core.Node, desiredVersion string) error {
+	err := drain.RunCordonOrUncordon(sc.drainer, node, false)
+	if err != nil {
+		return fmt.Errorf("unable to ensure node is uncordoned: %w", err)
+	}
+	// Version annotation is the indicator that the node was fully configured by this version of the services ConfigMap
+	if err = metadata.ApplyVersionAnnotation(sc.ctx, sc.client, *node, desiredVersion); err != nil {
```

Comment:
You need to apply the version annotation first and then uncordon the node.

Reply:
I disagree, as applying the version annotation is signaling to WMCO that the node has been completely been configured, and setting spec.unschedulable to false is part of that.

Reply:
I see your point. Would there be an issue if the version annotation does not get applied? i.e. the node start getting configured again just because of failure to update version?

Reply:
The node getting configured again in that would be the expected behavior, I can't think of any issues related to cordoning because of that. There would be a blip where the node is momentarily uncordoned and then recordened, as the controller goes to reconcile it again, but it should have no consequence. This would only be until the node is successfully configured.

Reply:
I feel we go by the k8s pattern a bit too much with retrying all operations just because the final operation failed.